File: python3/dsu.py
class DSU:
    def __init__(self):
        pass
    
    def init(self,n):
        self.pre=[i for i in range(n+1)]
    
    def find(self,x):
        if self.pre[x]!=x: self.pre[x]=self.find(self.pre[x])
        return self.pre[x]
    
    def merge(self,a,b):
        ra=self.find(a)
        rb=self.find(b)
        if ra!=rb:
            self.pre[ra]=rb
            return 1
        return 0

File: kotlin/dsu.kt
class Disjoint_Set_Union()
{
	lateinit var pre:IntArray
	lateinit var sz:IntArray
	fun init(n:Int)
	{
		pre=IntArray(n+1){it}
		sz=IntArray(n+1){1}
	}
	fun find(x:Int):Int
	{
		if(pre[x]!=x) pre[x]=find(pre[x])
		return pre[x]
	}
	fun merge(a:Int,b:Int):Boolean
	{
		var ra=find(a)
		var rb=find(b)
		if(ra!=rb)
		{
			pre[ra]=rb
			sz[rb]+=sz[ra]
			return true
		}
		return false
	}
}
